### PR TITLE
Refactor Closure

### PR DESCRIPTION
**Pull request by dsharletg**
This PR separates Closure from the LLVM-related bits, and moves GPU_Argument/GPU_Host_Closure to DeviceArgument/HostClosure to a header. This makes these tools more generally useful.

This does not have an immediate motivation, but it is part of a larger project that I'm trying to break into smaller, more frequent PRs.

### CODE REVIEW

File: src/DeviceArgument.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
+#ifndef HALIDE_DEVICE_ARGUMENT_H
+#define HALIDE_DEVICE_ARGUMENT_H
+
+/** \file
+ * Defines the code-generator interface for producing GPU device code
```
Comment (abadams) [inline src/DeviceArgument.h]: Is this comment right?

**Comment by abadams** [comment]
LGTM except for that file comment.

**Comment by zvookin** [comment]
LGTM. I do find the number of different variants of Argument types we have to be increasingly worrying. The issue being one of converting between them and not necessarily having all the information that one kind of argument thing has when it is needed in a different context. This is not reason enough to delay this CL, but longer term it would be nice to see if we can reduce the number of similar but a little bit different concepts in this area.

**Comment by dsharletg** [comment]
Note that this CL doesn't add a new one, it just renames GPU_Argument to DeviceArgument.

I tried to find a way to eliminate DeviceArgument and just use Argument. However, the GPU backends need a few extra parameters that are not easily eliminated. I think they *potentially* could be, by computing the information in DeviceArgument and not in Argument where it is consumed.

**Pull request merged**